source/extensions/common/wasm/wasm.cc

```diff
@@ -1016,6 +1037,153 @@ uint64_t Context::getCurrentTimeNanoseconds() {
+  using Filters::Common::Expr::CelValue;
+  switch (value.type()) {
+  case CelValue::Type::kMessage:
+    if (value.MessageOrDie() != nullptr) {
+      std::string result;
+      value.MessageOrDie()->SerializeToString(&result);
+      return result;
+    }
+  case CelValue::Type::kString:
+    return std::string(value.StringOrDie().value());
+  case CelValue::Type::kBytes:
+    return std::string(value.BytesOrDie().value());
+  case CelValue::Type::kInt64: {
+    const size_t n = sizeof(int64_t);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.Int64OrDie());
+    return std::string(out.buf, n);
+  }
+  case CelValue::Type::kUint64: {
+    const size_t n = sizeof(uint64_t);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.Uint64OrDie());
+    return std::string(out.buf, n);
+  }
+  case CelValue::Type::kDouble: {
+    const size_t n = sizeof(double);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.DoubleOrDie());
+    return std::string(out.buf, n);
+  }
+  case CelValue::Type::kBool: {
+    const size_t n = sizeof(bool);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.BoolOrDie());
+    return std::string(out.buf, n);
+  }
+  case CelValue::Type::kDuration: {
+    const size_t n = sizeof(absl::Duration);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.DurationOrDie());
+    return std::string(out.buf, n);
+  }
+  case CelValue::Type::kTimestamp: {
+    const size_t n = sizeof(absl::Time);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.TimestampOrDie());
+    return std::string(out.buf, n);
+  }
+  default:
+    // TODO: lists and maps
+    break;
+  }
+
+  return {};
+}
+
+absl::optional<std::string> Context::getSelectorExpression(absl::string_view path) {
+  using Filters::Common::Expr::CelValue;
+  using google::api::expr::runtime::FieldBackedListImpl;
+  using google::api::expr::runtime::FieldBackedMapImpl;
+
+  bool first = true;
+  CelValue value;
+  Protobuf::Arena arena;
+  const StreamInfo::StreamInfo& info = decoder_callbacks_->streamInfo();
+  const auto request_headers = request_headers_ ? request_headers_ : access_log_request_headers_;
+  const auto response_headers =
+      response_headers_ ? response_headers_ : access_log_response_headers_;
+  const auto response_trailers =
+      response_trailers_ ? response_trailers_ : access_log_response_trailers_;
+
+  size_t start = 0;
+  while (true) {
+    if (start >= path.size()) {
+      break;
+    }
+
+    size_t end = path.find('\0', start);
+    if (end == absl::string_view::npos) {
+      end = path.size();
+    }
+    auto part = path.substr(start, end - start);
+    start = end + 1;
+
+    // top-level ident
+    if (first) {
+      first = false;
+      if (part == "metadata") {
+        value = CelValue::CreateMessage(&info.dynamicMetadata(), &arena);
+      } else if (part == "request") {
+        value = CelValue::CreateMap(Protobuf::Arena::Create<Filters::Common::Expr::RequestWrapper>(
```

Comment:
Where is RequestWrapper defined?   Also isn't this very expensive: converting a structure into a proto just so we can use proto functions to access the values?

Reply:
It's in the base envoy. It's not a proto, it's a wrapper with no state.

Reply:
The code might look weird since I could not find a way to optionally stack allocate the wrapper.

Reply:
Ah, I see, that code is from Aug 19th, no wonder I couldn't find it.

Just curious, what happens when you try to stack allocate it.  It seems to have a standard constructor.

Reply:
Doesn't the body of the loop correspond to a code block? The compiler is free to delete the stack allocated value here, right? It might work, just worried whether that is introducing some undefined behavior.

Reply:
I see what you are saying. You don't want to build the wrappers outside the loop because you aren't sure you are going to need them all.  This is a very reasonable solution.  The alternative of using alloca is not appealing in a loop.